pkg/controller/metrics.go

```diff
@@ -1,58 +1,163 @@
-			Subsystem: subsystem,
-			Name:      "errors_total",
-			Help:      "The total number or unexpected errors the controller encountered",
+			Name: metricsPrefix + "errors_total",
+			Help: "The total number or unexpected errors the controller encountered",
 		}),
-		ControllerOperation: prometheus.NewHistogramVec(prometheus.HistogramOpts{
-			Namespace: namespace,
-			Subsystem: subsystem,
-			Name:      "controller_operation_duration_seconds",
-			Help:      "The duration it takes to execute an operation",
-		}, []string{"operation"}),
-		NodeJoinDuration: prometheus.NewHistogramVec(prometheus.HistogramOpts{
-			Namespace: namespace,
-			Subsystem: subsystem,
-			Name:      "node_join_duration_seconds",
-			Help:      "The time it takes from creation of the machine resource and the final creation of the node resource",
-		}, []string{}),
 	}
 
 	// Set default values, so that these metrics always show up
-	cm.Machines.Set(0)
 	cm.Workers.Set(0)
-	cm.Nodes.Set(0)
+	cm.Errors.Add(0)
 
 	return cm
 }
+
+type MachineCollector struct {
+	lister v1alpha1.MachineLister
+
+	machines       *prometheus.Desc
+	machineCreated *prometheus.Desc
+	machineDeleted *prometheus.Desc
+}
+
+func NewMachineCollector(lister v1alpha1.MachineLister) *MachineCollector {
+	return &MachineCollector{
+		lister: lister,
+
+		machines: prometheus.NewDesc(
+			metricsPrefix+"machines",
+			"The number of machines managed by this machine controller",
+			nil, nil,
+		),
+		machineCreated: prometheus.NewDesc(
+			metricsPrefix+"machine_created",
+			"Timestamp of the machine's creation time",
+			[]string{"machine"}, nil,
+		),
+		machineDeleted: prometheus.NewDesc(
+			metricsPrefix+"machine_deleted",
+			"Timestamp of the machine's deletion time",
+			[]string{"machine"}, nil,
+		),
+	}
+}
+
+func (mc MachineCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- mc.machines
+	ch <- mc.machineCreated
+	ch <- mc.machineDeleted
+}
+
+func (mc MachineCollector) Collect(ch chan<- prometheus.Metric) {
+	machines, err := mc.lister.List(labels.Everything())
+	if err != nil {
+		return
+	}
+
+	ch <- prometheus.MustNewConstMetric(
+		mc.machines,
+		prometheus.GaugeValue,
+		float64(len(machines)),
+	)
+
+	for _, machine := range machines {
+		ch <- prometheus.MustNewConstMetric(
+			mc.machineCreated,
+			prometheus.GaugeValue,
+			float64(machine.CreationTimestamp.Unix()),
+			machine.Name,
+		)
+
+		if machine.DeletionTimestamp != nil {
+			ch <- prometheus.MustNewConstMetric(
+				mc.machineDeleted,
+				prometheus.GaugeValue,
+				float64(machine.DeletionTimestamp.Unix()),
+				machine.Name,
+			)
+		}
+	}
+}
+
+type NodeController struct {
+	lister v1.NodeLister
+
+	nodes       *prometheus.Desc
+	nodeCreated *prometheus.Desc
+	nodeDeleted *prometheus.Desc
+}
+
+func NewNodeController(lister v1.NodeLister) *NodeController {
+	return &NodeController{
+		lister: lister,
+
+		nodes: prometheus.NewDesc(
+			metricsPrefix+"nodes",
+			"The number of nodes created by a machine",
```

Comment:
I am actually not sure if this metric makes sense, because it will include __all__ nodes, not only the ones that were created by a machine-controller, we could get the same info from kube state metrics

Reply:
Yup, had the same thought yesterday. Then I also looked into kube-state-metrics again and found `kube_node_created`. 
The reason I added them back were 
1. We already had node metrics.
2. You still get node metrics, even though you have not deployed kube-state-metrics.

I'm fine with removing them.